### PR TITLE
Add the Label attribute

### PR DESCRIPTION
**Pull request by dood-**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #633

### CODE REVIEW

File: src/Helper/Label.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Helper;
```
Comment (vjik) [inline src/Helper/Label.php:5]: Suggest move attribute `Label` to root (`Yiisoft\Validator`). It's not helper.

File: src/LabelsProviderInterface.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Provides data attribute labels.
+ */
+interface LabelsProviderInterface
+{
+    /**
+     * @return array<string, string> A set of attribute labels.
+     */
+    public function getLabels(): array;
```
Comment (vjik) [inline src/LabelsProviderInterface.php:15]:
```suggestion
    public function getValidationPropertyLabels(): array;
```

`getLabels` it's more common name and may lead to conflicts in some cases.

**Review by vjik — CHANGES_REQUESTED** [review]
Need test when validated object implements `LabelsProviderInterface` with and without `Label` attributes usage.

File: src/DataSet/ObjectDataSet.php [hunk 3]
```diff
@@ -158,6 +159,10 @@ final class ObjectDataSet implements RulesProviderInterface, DataWrapperInterfac
      * @var bool Whether an {@see $object} provided rules by implementing {@see RulesProviderInterface}.
      */
     private bool $rulesProvided;
+    /**
+     * @var bool Whether an {@see $object} provided rules by implementing {@see LabelsProviderInterface}.
```
Comment (vjik) [inline src/DataSet/ObjectDataSet.php:163]:
```suggestion
     * @var bool Whether an {@see $object} provided labels by implementing {@see LabelsProviderInterface}.
```

File: tests/Support/Data/ObjectWithLabelsProvider.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+final class ObjectWithLabelsProvider implements LabelsProviderInterface
+{
+    #[Required(message: '{attribute} cannot be blank.')]
+    public string $name = '';
+
+    #[Number(min: 21, lessThanMinMessage: '{attribute} must be no less than {min}.')]
+    protected int $age = 17;
+
+    #[Number(max: 100)]
+    #[Label('test')]
```
Comment (vjik) [inline tests/Support/Data/ObjectWithLabelsProvider.php:21]: Need additional test for case when propert contain attribute `Label` and `getValidationPropertyLabels()` contain label for same property.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/DataSet/ObjectDataSet.php [hunk 5]
```diff
@@ -158,6 +159,10 @@ final class ObjectDataSet implements RulesProviderInterface, DataWrapperInterfac
      * @var bool Whether an {@see $object} provided rules by implementing {@see RulesProviderInterface}.
      */
     private bool $rulesProvided;
+    /**
+     * @var bool Whether an {@see $object} provided labels by implementing {@see LabelsProviderInterface}.
+     */
+    private bool $labelsProvided;
```
Comment (samdark) [inline src/DataSet/ObjectDataSet.php:165]: No need to cache `instanceof` in a boolean variable, it is fast enough to check directly in `getValidationPropertyLabels()`.

File: src/DataSet/ObjectDataSet.php [hunk 6]
```diff
@@ -299,4 +305,15 @@ public function getAttributeTranslator(): ?AttributeTranslatorInterface
     {
         return $this->parser->getAttributeTranslator();
     }
+
+    public function getValidationPropertyLabels(): array
```
Comment (samdark) [inline src/DataSet/ObjectDataSet.php:309]: How about shorter name?

```suggestion
    public function getPropertyLabels(): array
```

File: src/LabelsProviderInterface.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Provides data attribute labels.
+ */
+interface LabelsProviderInterface
+{
+    /**
+     * @return array<string, string> A set of attribute labels.
+     */
+    public function getValidationPropertyLabels(): array;
```
Comment (samdark) [inline src/LabelsProviderInterface.php:15]:
```suggestion
    public function getPropertyLabels(): array;
```

Comment (samdark) [reply src/LabelsProviderInterface.php:15]: Or was the name chosen not to conflict with class method names?

File: src/ValidationContext.php [hunk 8]
```diff
@@ -227,6 +235,14 @@ public function getAttribute(): ?string
         return $this->attribute;
     }
 
+    /**
+     * @return string|null
+     */
+    public function getAttributeLabel(): ?string
```
Comment (samdark) [inline src/ValidationContext.php:241]: Not covered by tests. Also, phpdoc is redundant.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
👍 overall.

File: src/Validator.php [hunk 9]
```diff
@@ -131,6 +131,10 @@ public function validate(
                 $validatedData = $dataSet->getAttributeValue($attribute);
                 $context->setParameter(ValidationContext::PARAMETER_VALUE_AS_ARRAY, null);
                 $context->setAttribute($attribute);
+
+                if ($dataSet instanceof LabelsProviderInterface) {
+                    $context->setAttributeLabel($dataSet->getValidationPropertyLabels()[$attribute] ?? null);
```
Comment (arogachev) [inline src/Validator.php:136]: Shouldn't we add:

```php
$context->setAttributeLabel(null);
```
in the other part of condition similar to:

https://github.com/yiisoft/validator/blob/163a813ac1d5b7ba61949c9eed4e09c80c954038/src/Validator.php#L128

?

Not sure about this one, but it's related to `Nested` and `Each` rules combination.

File: src/LabelsProviderInterface.php (hunk 7 above)
Comment (dood-) [reply src/LabelsProviderInterface.php:15]: Yes, @vjik suggested replacing `getLabels` with `getValidationPropertyLabels`

File: tests/Helper/ObjectParserTest.php [hunk 10]
```diff
@@ -116,14 +116,18 @@ public function testCache(): void
...
+        $labels1 = $parser1->getLabels();
         $this->assertEquals($expectedRules1, $rules1);
         $cache = $cacheProperty->getValue();
         $this->assertArrayHasKey($cacheKey1, $cache);
         $this->assertArrayHasKey('rules', $cache[$cacheKey1]);
         $this->assertArrayHasKey('reflectionProperties', $cache[$cacheKey1]);
         $this->assertArrayHasKey('reflectionSource', $cache[$cacheKey1]);
+        $this->assertArrayHasKey('labels', $cache[$cacheKey1]);
         $this->assertSame($rules1, $parser1->getRules());
+        $this->assertEquals($labels1,  $expectedLabels1);
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:130]:
```suggestion
        $this->assertSame($labels1,  $expectedLabels1);
```

File: tests/Helper/ObjectParserTest.php [hunk 11]
```diff
@@ -170,10 +174,14 @@ public function testDisabledCache(): void
             'b' => [new Number(min: 1)],
             'c' => [new Number(max: 2)],
         ];
+        $expectedLabels = ['c' => 'label'];
         $rules = $parser->getRules();
+        $labels = $parser->getLabels();
         $this->assertEquals($expectedRules, $rules);
+        $this->assertEquals($expectedLabels, $labels);
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:181]:
```suggestion
        $this->assertSame($expectedLabels, $labels);
```

File: tests/Helper/ObjectParserTest.php [hunk 12]
```diff
@@ -170,10 +174,14 @@ public function testDisabledCache(): void
...
         $this->assertArrayNotHasKey($cacheKey, $cacheProperty->getValue());
         $this->assertEquals($expectedRules, $parser->getRules());
+        $this->assertEquals($expectedLabels, $parser->getLabels());
```
Comment (arogachev) [inline tests/Helper/ObjectParserTest.php:184]:
```suggestion
        $this->assertSame($expectedLabels, $parser->getLabels());
```

File: tests/ValidatorTest.php [hunk 13]
```diff
@@ -1415,4 +1417,42 @@ public function testDifferentValueAsArrayInSameContext(): void
         );
         $this->assertTrue($result->isValid());
     }
+
+
+    public function dataDataWithLabels(): array
```
Comment (arogachev) [inline tests/ValidatorTest.php:1422]:
```suggestion
    public function dataErrorMessagesWithLabels(): array
```

File: tests/ValidatorTest.php [hunk 14]
```diff
@@ -1415,4 +1417,42 @@ public function testDifferentValueAsArrayInSameContext(): void
...
+                [
+                    'name' => ['Имя cannot be blank.'],
+                    'age' => ['Возраст must be no less than 21.'],
+                ],
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider dataDataWithLabels
```
Comment (arogachev) [inline tests/ValidatorTest.php:1448]:
```suggestion
     * @dataProvider dataErrorMessagesWithLabels
```

**Review by arogachev — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

File: src/Validator.php [hunk 15]
```diff
@@ -131,6 +131,10 @@ public function validate(
...
+                }
```
Comment (vjik) [inline src/Validator.php:137]: Seems, currently may be case when label will be incorrect (keep from previous context). 

I think, label setting should be here:

image

**Review by vjik — CHANGES_REQUESTED** [review]

**Review by vjik — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**